src/integrators/rosenbrock_integrators.jl

```diff
@@ -1,335 +1,387 @@
-function initialize!(integrator,cache::Rosenbrock23Cache,f=integrator.f)
+function initialize!(integrator, cache::Rosenbrock23Cache)
   integrator.kshortsize = 2
   @unpack k₁,k₂,fsalfirst,fsallast = cache
   integrator.fsalfirst = fsalfirst
   integrator.fsallast = fsallast
   integrator.k = [k₁,k₂]
-  f(integrator.t,integrator.uprev,integrator.fsalfirst)
+  integrator.f(integrator.t, integrator.uprev, integrator.fsalfirst)
 end
 
-@muladd function perform_step!(integrator,cache::Rosenbrock23Cache,f=integrator.f)
-  @unpack t,dt,uprev,u = integrator
-  uidx = eachindex(integrator.uprev)
+@muladd function perform_step!(integrator, cache::Rosenbrock23Cache, repeat_step=false)
+  @unpack t,dt,uprev,u,f = integrator
   @unpack k₁,k₂,k₃,du1,du2,f₁,vectmp,vectmp2,vectmp3,fsalfirst,fsallast,dT,J,W,tmp,uf,tf,linsolve_tmp,linsolve_tmp_vec,jac_config = cache
   @unpack c₃₂,d = cache.tab
-  mass_matrix = integrator.sol.prob.mass_matrix
 
-  # Setup Jacobian Calc
+  # Assignments
   sizeu  = size(u)
-  tf.vf.sizeu = sizeu
-  tf.uprev = uprev
-  uf.vfr.sizeu = sizeu
-  uf.t = t
+  mass_matrix = integrator.sol.prob.mass_matrix
 
-  if has_tgrad(f)
-    f(Val{:tgrad},t,uprev,dT)
-  else
-    if alg_autodiff(integrator.alg)
-      ForwardDiff.derivative!(dT,tf,vec(du2),t) # Should update to inplace, https://github.com/JuliaDiff/ForwardDiff.jl/pull/219
+  # Precalculations
+  γ = dt*d
+  dto2 = dt/2
+  dto6 = dt/6
+
+  # Time derivative
+  if !repeat_step # skip calculation if step is repeated
+    if has_tgrad(f)
+      f(Val{:tgrad}, t, uprev, dT)
     else
-      dT = Calculus.finite_difference(tf,t,integrator.alg.diff_type)
+      tf.vf.sizeu = sizeu
+      tf.uprev = uprev
+      if alg_autodiff(integrator.alg)
+        ForwardDiff.derivative!(dT, tf, vec(du2), t)
+      else
+        dT = Calculus.finite_difference(tf, t, integrator.alg.diff_type)
+      end
     end
   end
 
-  γ = dt*d
-
   @. linsolve_tmp = fsalfirst + γ*dT
 
+  # Jacobian
   if has_invW(f)
-    f(Val{:invW},t,uprev,γ,W) # W == inverse W
-    A_mul_B!(vectmp,W,linsolve_tmp_vec)
+    # skip calculation of inv(W) if step is repeated
+    repeat_step || f(Val{:invW}, t, uprev, γ, W) # W == inverse W
```

Comment:
Same as `!repeat_step &&`? I find this a little non-standard and would like to have it using `&&` instead since that's easier for me to read.